### PR TITLE
DrivAerML: AdamW beta2=0.99/0.995 faster adaptation

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PyTorch AdamW defaults to beta2=0.999, which means the second-moment estimate adapts over ~1000 steps. DrivAerML has a non-stationary loss landscape: cosine LR cycling (T_max=30) creates periodic gradient magnitude changes, and batch_size=1 with random surface-point sampling creates high variance. Faster second-moment adaptation (beta2=0.99, adapting over ~100 steps) may better track these rapid changes and find deeper basins.

This is a completely unexplored hyperparameter axis for DrivAerML. jet (#3028) was assigned a beta sweep but is now on a different experiment.

## Instructions

Run two variants — beta2=0.99 and beta2=0.995 — at the exact champion config:

**Run 1: beta2=0.99 (faster adaptation)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --adam-beta2 0.99 \
  --wandb_group dm-adamw-beta-sweep
```

**Run 2: beta2=0.995 (intermediate)**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --adam-beta2 0.995 \
  --wandb_group dm-adamw-beta-sweep
```

**Implementation:** Add `--adam-beta2` flag to train.py. When building the AdamW optimizer, pass `betas=(0.9, args.adam_beta2)` instead of the default `(0.9, 0.999)`.

**Divergence check:** If either run diverges before epoch 30, add `--grad-clip 0.5` and re-run.

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch for each variant
- Test `surface_rel_l2_pct` at best val checkpoint
- W&B run IDs

## Baseline

**DrivAerML champion (val):**
- `val_primary/surface_rel_l2_pct`: **3.997%** at epoch 467
- W&B run: `bht6h42t` (PR #2898, 4L/512d, AdamW lr=5e-4, T_max=30, betas default (0.9, 0.999))
- External targets: Transolver-3 3.71% | AB-UPT 3.82%

```bash
# Champion reproduce
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT einar:

## Results

### Summary

**Both beta2 variants underperform the baseline and eventually diverge.** Lower beta2 provides faster second-moment adaptation but makes training unstable with cosine annealing restarts — the optimizer overreacts to gradient magnitude changes at restart boundaries.

### Metrics

| Variant | beta2 | Best Val Epoch | Val `surface_rel_l2_pct` | Test `surface_rel_l2_pct` | Diverged At | W&B Run |
|---------|-------|---------------|--------------------------|---------------------------|-------------|---------|
| **Baseline** | 0.999 | 467 | **3.997%** | — | No | `bht6h42t` |
| Run 1 | 0.99 | 340 | 4.431% | 5.569% | Epoch 413 (NaN) | `ufbx7vfj` |
| Run 2 | 0.995 | 105 | 7.653% | 8.599% | Epoch 114 (43.7%) | `jedieqgl` |

External targets: Transolver-3 3.71% | AB-UPT 3.82%

### Commands

```bash
# Run 1: beta2=0.99
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --adam-beta2 0.99 --agent einar --wandb-name "einar/dm-beta2-099" --wandb-group dm-adamw-beta-sweep

# Run 2: beta2=0.995
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --adam-beta2 0.995 --agent einar --wandb-name "einar/dm-beta2-0995" --wandb-group dm-adamw-beta-sweep
```

### What happened

**Negative result.** Both lower beta2 variants performed worse than the default beta2=0.999 and eventually diverged:

- **beta2=0.99** converged to 4.431% val (10.8% worse than baseline 3.997%) by epoch 340, then diverged to NaN at epoch 413. The faster adaptation (~100-step window) initially tracks gradient changes well but becomes unstable during later cosine restarts.

- **beta2=0.995** performed much worse — best val only 7.653% at epoch 105, then diverged catastrophically at epoch 114 with gradient norms exploding to trillions. The intermediate adaptation rate hit a bad resonance with the cosine restart period.

The default beta2=0.999 (~1000-step adaptation) appears to be the right timescale for DrivAerML with T_max=30 cosine annealing. The slower adaptation smooths over the periodic gradient changes at restart boundaries rather than amplifying them.

### Suggested follow-ups

- **beta2=0.9995 or 0.9999** — if anything, the results suggest *slower* second-moment adaptation could help stability in late training. Worth testing if the current champion eventually diverges on longer runs.
- **beta2 scheduling** — start with beta2=0.99 for faster early convergence, then anneal to 0.999 for stability. This could combine the fast early learning of beta2=0.99 (which was tracking the baseline until epoch ~300) with the late-training stability of the default.
- **Grad clipping + lower beta2** — the divergence might be preventable with gradient clipping. A beta2=0.99 + grad_clip=0.5 combination could recover the convergence speed without the instability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3110 because ADVISOR: Closing — both beta2 variants (0.99, 0.995) underperform and diverge. beta2=0.999 (default) is correct for DM. Einar reassigned..
